### DISCUSSION TITLE
Automatic commenting on a PR

### DISCUSSION THREAD

**Discussion by sayakpaul** [community/community#26002] [Actions]
<p>Hi folks.</p>
<p>I am trying to extend what I described <a href="https://github.community/t/automatic-commenting-on-a-commit-with-results-from-a-script">here</a> to automatic commenting on the latest PR I would create.</p>
<p>Here’s what I have done so far -</p>
<ul>
<li>Created <a href="https://github.com/sayakpaul/wine/blob/experiment-1/create_pr_comment.py" rel="nofollow noopener">this script</a> that would create the comment with the PR and Issues GitHub APIs.</li>
<li>Wrote <a href="https://github.com/sayakpaul/wine/blob/experiment-1/.github/workflows/pr.yaml" rel="nofollow noopener">this sample workflow</a> that would execute that script to facilitate the commenting.</li>
</ul>
<p>With this setup, the <code>create_pr_comment.py</code> script is failing to get executed. -</p>
<p><img src="https://user-images.githubusercontent.com/22957388/181094711-2b59e64a-7786-4ef0-921f-1cba50ad670d.png" alt="image" data-base62-sha1="fmFWfRvh5N3EDilDzwrxva3oq9Y" width="464" height="341"></p>
<p>It’s very weird to me that <code>requests</code> is unavailable in standard Python setup. Or am I missing something out?</p>
<p><strong>Edit</strong>: I was able to resolve the issue and I can now see automatic comments on the latest PR -</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/a/acf94c64a5a88259086f3dd35ecea2ba4c3786d6.png" data-download-href="/uploads/short-url/oGcf95hgxIhAIm5kBxzDNSf7V9c.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/22957388/181094712-aa13d0f2-1cbf-4fff-ada7-f525fbc620a3.png" alt="image" data-base62-sha1="oGcf95hgxIhAIm5kBxzDNSf7V9c" width="690" height="87" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/a/acf94c64a5a88259086f3dd35ecea2ba4c3786d6_2_690x87.png, https://aws1.discourse-cdn.com/github/original/2X/a/acf94c64a5a88259086f3dd35ecea2ba4c3786d6.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/a/acf94c64a5a88259086f3dd35ecea2ba4c3786d6.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/a/acf94c64a5a88259086f3dd35ecea2ba4c3786d6_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">949×120 14.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Is it possible to have a GitHub bot make this comment? I tried the following, but it did not work out -</p>
<pre><code class="lang-auto">pip install requests
git config --global user.name github-actions[bot]
git config --global user.email 41898282+github-actions[bot]@users.noreply.github.com
python create_pr_comment.py
</code></pre>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="sayakpaul" data-post="1" data-topic="130978">
<div class="title">
<div class="quote-controls"></div>
 sayakpaul:</div>
<blockquote>
<p>Is it possible to have a GitHub bot make this comment?</p>
</blockquote>
</aside>
<p>According to <a href="https://docs.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#permissions-for-the-github_token">Permissions for the <code>GITHUB_TOKEN</code> </a> documentation the <code>GITHUB_TOKEN</code> has write permission on issues, so you should be able to use that instead of your PAT. The git config doesn’t have anything to do with the Github API.</p>
<p>Two other things:</p>
<ul>
<li>
<p>Depending on who’s pull requests your workflow should comment on, you might want to use <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request_target"><code>pull_request_target</code></a> instead of <code>pull_request</code> to trigger your workflow. For <code>pull_request</code> events from forks the rights associated with the <code>GITHUB_TOKEN</code> are heavily restricted and secrets unavailable, because the workflow itself could have been modified, possibly maliciously.</p>
</li>
<li>
<p>You’re passing the token to your script on the command line. This is a security risk: If someone could check the running processes on the runner (think <code>ps xau</code> or similar) they would be able to see and copy your token. With <code>GITHUB_TOKEN</code> impact would be somewhat limited, but a PAT might be exploited for a long time, if the attack isn’t noticed and the token revoked. Using <code>env</code> would be safer.</p>
</li>
</ul>

**Comment by sayakpaul** [comment]
<p>Thanks, <a class="mention" href="/airtower-luna">@airtower-luna</a>. I incorporated your suggestions.  But I <a href="https://github.com/sayakpaul/wine/actions/runs/243310602" rel="nofollow noopener">do not see any workflow to trigger</a>. I am probably missing out on something.</p>
<p>Also, could you elaborate on what did you mean by <code>env</code>?</p>
<aside class="quote no-group" data-username="airtower-luna" data-post="2" data-topic="130978">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>Using <code>env</code> would be safer.</p>
</blockquote>
</aside>

**Comment by airtower-luna** [comment]
<p>The linked workflow doesn’t seem to have anything to do with the one discussed above. :sweat_smile:</p>
<aside class="quote no-group" data-username="sayakpaul" data-post="3" data-topic="130978">
<div class="title">
<div class="quote-controls"></div>
 sayakpaul:</div>
<blockquote>
<p>Also, could you elaborate on what did you mean by <code>env</code> ?</p>
</blockquote>
</aside>
<p>Use an environment variable to pass the token to your script, like so:</p>
<pre><code class="lang-yaml">- run: python create_pr_comment.py
  env:
    TOKEN: ${{ secrets.GITHUB_TOKEN }}
</code></pre>
<p>And in your script use</p>
<pre><code class="lang-python">os.environ['TOKEN']
</code></pre>
<p>to get the value. This has the advantage that the token isn’t part of the command line, where it can easily be seen by other processes. The risk that someone else is able to run commands on the runner VM is probably low, but its good to always keep this kind of thing in mind.</p>

**Comment by sayakpaul** [comment]
<aside class="quote no-group" data-username="airtower-luna" data-post="4" data-topic="130978">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>The linked workflow doesn’t seem to have anything to do with the one discussed above. :sweat_smile:</p>
</blockquote>
</aside>
<p>This is because <a href="https://github.com/sayakpaul/wine/blob/d5108291759257c3f183c354a5e4f5c6d9454e13/.github/workflows/pr.yaml" rel="nofollow noopener">this workflow</a> never ran.</p>

**Comment by kingthorin** [comment]
<p>Here’s one you can model from:<br>
</p><aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/OWASP/www-project-vulnerable-web-applications-directory/blob/94c5605064375de63e9e266bcf5a4cba08fa2193/.github/workflows/pr_comment.yml" target="_blank" rel="nofollow noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/OWASP/www-project-vulnerable-web-applications-directory/blob/94c5605064375de63e9e266bcf5a4cba08fa2193/.github/workflows/pr_comment.yml" target="_blank" rel="nofollow noopener">OWASP/www-project-vulnerable-web-applications-directory/blob/94c5605064375de63e9e266bcf5a4cba08fa2193/.github/workflows/pr_comment.yml</a></h4>
<pre><code class="lang-yml">name: PR Comment

on:
  pull_request_target:
    types:
      - opened
    paths:
    - '**.json'

jobs:
  PR-Comment:
    if: github.actor != 'kingthorin' &amp;&amp; github.actor != 'vwadbot' &amp;&amp; github.actor != 'psiinon' &amp;&amp; github.actor != 'raulsiles'  
    runs-on: ubuntu-latest
    steps:
    - name: PR Comment
      uses: actions/github-script@v2
      with:
        github-token: ${{secrets.GITHUB_TOKEN}}
        script: |
          github.issues.createComment({
</code></pre>

  This file has been truncated. <a href="https://github.com/OWASP/www-project-vulnerable-web-applications-directory/blob/94c5605064375de63e9e266bcf5a4cba08fa2193/.github/workflows/pr_comment.yml" target="_blank" rel="nofollow noopener">show original</a>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Here’s an example of it in action:<br>
</p><aside class="onebox githubpullrequest">
  <header class="source">
      <a href="https://github.com/OWASP/www-project-vulnerable-web-applications-directory/pull/18" target="_blank" rel="nofollow noopener">github.com/OWASP/www-project-vulnerable-web-applications-directory</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Pull Request">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 12 16" aria-hidden="true"><path d="M11 11.28V5c-.03-.78-.34-1.47-.94-2.06C9.46 2.35 8.78 2.03 8 2H7V0L4 3l3 3V4h1c.27.02.48.11.69.31.21.2.3.42.31.69v6.28A1.993 1.993 0 0 0 10 15a1.993 1.993 0 0 0 1-3.72zm-1 2.92c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zM4 3c0-1.11-.89-2-2-2a1.993 1.993 0 0 0-1 3.72v6.56A1.993 1.993 0 0 0 2 15a1.993 1.993 0 0 0 1-3.72V4.72c.59-.34 1-.98 1-1.72zm-.8 10c0 .66-.55 1.2-1.2 1.2-.65 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2zM2 4.2C1.34 4.2.8 3.65.8 3c0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/OWASP/www-project-vulnerable-web-applications-directory/pull/18" target="_blank" rel="nofollow noopener">Adding a new online resource</a>
    </h4>

    <div class="branches">
      <code>OWASP:master</code> ← <code>arthurakay:patch-1</code>
    </div>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2020-08-24" data-time="15:49:15" data-timezone="UTC">03:49PM - 24 Aug 20 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/arthurakay" target="_blank" rel="nofollow noopener">
          <img alt="arthurakay" src="https://user-images.githubusercontent.com/7570458/181094715-d20e8fac-8b92-4e85-8edf-0df6acf74253.png" class="onebox-avatar-inline" width="20" height="20">
          arthurakay
        </a>
      </div>

      <div class="lines" title="1 commits changed 1 files with 17 additions and 0 deletions">
        <a href="https://github.com/OWASP/www-project-vulnerable-web-applications-directory/pull/18/files" target="_blank" rel="nofollow noopener">
          <span class="added">+17</span>
          <span class="removed">-0</span>
        </a>
      </div>
    </div>

  </div>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by sayakpaul** [comment]
<p>Thank you for chiming in. But currently, I would like to use <a href="https://github.com/sayakpaul/wine/blob/17ac2ef10e65e6926b415d4b5d0a0ed3a83f68bc/create_pr_comment.py" rel="nofollow noopener">this script</a> when <code>pull_request_target</code> is triggered.</p>

**Comment by sayakpaul** [comment]
<p>I <a href="https://github.com/sayakpaul/wine/blob/ffc8f6cb611707928475bdb670c966afba65a019/.github/workflows/pr_new.yaml" rel="nofollow noopener">did try it</a> but it did not run. Maybe I am missing out on something?</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a>,</p>
<p>For event ‘pull_request_target’,<br>
<img src="https://user-images.githubusercontent.com/52772761/181094717-0683ccb0-6f7b-42eb-b8b5-b87d168e751a.png" alt="image" data-base62-sha1="sFRhHyCB67GS3QQq2Zxs0pRnvEu" width="405" height="87"></p>
<p>Which means <code>actions/checkout</code> will get codes from <strong>base branch</strong> by default, not the fake merge branch or PR compare branch.</p>
<p>If your code exists in PR compare branch, you need to specify the repo(and may be with ref), such as:</p>
<pre><code class="lang-auto">- name: Checkout tools repo
  uses: actions/checkout@v2
  with:
    repository: weide-zhou/ticket24
    ref: dev
</code></pre>
<p>And i also tried the <a href="https://github.com/sayakpaul/wine/blob/ffc8f6cb611707928475bdb670c966afba65a019/.github/workflows/pr_new.yaml" rel="nofollow noopener">workflow</a> same as you, it works fine on my side.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/f/fef81e04ba2601c6e069ade7e1001ea94e0882f7.png" data-download-href="/uploads/short-url/AnyRHWFFXeWzx8EIylVVVrPUx2D.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181094718-d7f3d142-81fc-4822-b9f8-618a98b2b717.png" alt="image" data-base62-sha1="AnyRHWFFXeWzx8EIylVVVrPUx2D" width="690" height="319" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/f/fef81e04ba2601c6e069ade7e1001ea94e0882f7_2_690x319.png, https://aws1.discourse-cdn.com/github/original/2X/f/fef81e04ba2601c6e069ade7e1001ea94e0882f7.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/f/fef81e04ba2601c6e069ade7e1001ea94e0882f7.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/f/fef81e04ba2601c6e069ade7e1001ea94e0882f7_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">850×394 23.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>My workflow is <a href="https://github.com/weide-zhou/ticket34/runs/1084541430?check_suite_focus=true" rel="nofollow noopener">here</a>. Please set the event ‘pull_request_target’ on your <strong>base repo and base branch</strong>, not the PR compare branch.</p>
<p>Thanks</p>

**Comment by kingthorin** [comment]
<aside class="quote no-group" data-username="sayakpaul" data-post="8" data-topic="130978" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 sayakpaul:</div>
<blockquote>
<p>I <a href="https://github.com/sayakpaul/wine/blob/ffc8f6cb611707928475bdb670c966afba65a019/.github/workflows/pr_new.yaml" rel="nofollow noopener">did try it </a> but it did not run. Maybe I am missing out on something?</p>
</blockquote>
</aside>
<p>It won’t run in the PR that’s adding it. (This seems to be a discrepancy of <code>pull_request_target</code> vs <code>pull_request</code> workflows). Nor will it run against PRs that don’t have the workflow in their base state/branch  (ex: when the “feature” branch is created from master [or whatever default] before it contained the workflow to be executed.)</p>

**Comment by sayakpaul** [comment]
<p>Thanks <a class="mention" href="/kingthorin">@kingthorin</a> for the explanation.  In my case, what would you suggest?</p>

**Comment by sayakpaul** [comment]
<p>Thanks so much for the help. <a class="mention" href="/weide-zhou">@weide-zhou</a>.</p>
<p>Quick question: I modified <a href="https://github.com/sayakpaul/wine/blob/experiment-1/.github/workflows/pr.yaml" rel="nofollow noopener"><code>pr.yaml</code></a> to use the following -</p>
<pre><code class="lang-auto">- uses: actions/checkout@v2
        with:
          repository: sayakpaul/wine
          ref: experiment-1
</code></pre>
<p>But the workflow still does not run it appears. Anything I am missing out on?</p>

**Comment by kingthorin** [comment]
<p>If the action/github-script based solution works for your then commit it directly to your default branch (or create a PR and merge it) then open another PR to see that it works…</p>

**Comment by sayakpaul** [comment]
<p>How would I pass an environmental variable to the <code>body</code> parameter of <code> github.issues.createComment</code>?</p>
<p>I have loaded some content from a file into a variable and I would like to pass that to <code>body</code>. Here’s the workflow file for your perusal.</p>
<pre><code class="lang-auto">name: PR Comment

on:
  pull_request:

jobs:
  PR-Comment:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: Get run URL
      run: |
        body=$(&lt;run.txt)
    - name: PR Comment
      uses: actions/github-script@v2
      with:
        github-token: ${{secrets.GITHUB_TOKEN}}
        script: |
          github.issues.createComment({
            issue_number: ${{ github.event.number }},
            owner: context.repo.owner,
            repo: context.repo.repo,
            body: body
          })
</code></pre>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a>,</p>
<p>Your action cannot get the var $body since they are in separated steps.<br>
You can get the run.txt content and set it as env, and send it to the action as parameter. Fixed as below:</p>
<pre><code class="lang-auto">  PR-Comment:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: get the run URL
      run: |
        echo "::set-env name=body::$(&lt;run.txt)"   # set it as env
    - name: PR Comment
      uses: actions/github-script@v2
      with:
        github-token: ${{secrets.GITHUB_TOKEN}}
        script: |
          github.issues.createComment({
            issue_number: ${{ github.event.number }},
            owner: context.repo.owner,
            repo: context.repo.repo,
            body: '${{env.body}}'   # use '${{env.xxx}}' as the parameter
          })
</code></pre>
<p>Thanks</p>

**Comment by sayakpaul** [comment]
<p>Another useful lesson. Thank you so much.</p>
